### PR TITLE
Spring Boot practice: Add ShoppingCart

### PR DESCRIPTION
**Pull request by Chorniypistolet**

### CODE REVIEW

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+
+@RestController
+@RequestMapping("/cart")
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    ShoppingCartDto addBookToCart(@AuthenticationPrincipal User user, @RequestBody CartItemRequestDto cartItemRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto addBookToCart(@AuthenticationPrincipal User user, @RequestBody @Valid CartItemRequestDto cartItemRequestDto) {
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        return shoppingCartService.addBookToCart(user.getId(), cartItemRequestDto);
+    }
+    @GetMapping
+    public ShoppingCartDto getShoppingCart (@AuthenticationPrincipal User user){
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto getShoppingCart(@AuthenticationPrincipal User user){
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+@RestController
+@RequestMapping("/cart")
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    ShoppingCartDto addBookToCart(@AuthenticationPrincipal User user, @RequestBody CartItemRequestDto cartItemRequestDto) {
+        return shoppingCartService.addBookToCart(user.getId(), cartItemRequestDto);
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java:28]:
```suggestion
    }
    
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    @GetMapping
+    public ShoppingCartDto getShoppingCart (@AuthenticationPrincipal User user){
+        return shoppingCartService.getCartByUserId(user.getId());
+    }
+
+    @PutMapping("/items/{id}")
+    public ShoppingCartDto updateCartItemQuantity (@PathVariable("id") Long id,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemQuantity(@PathVariable("id") Long id,
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            @RequestBody CartItemRequestDto cartItemRequestDto){
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
            @RequestBody @Valid CartItemRequestDto cartItemRequestDto){
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        return shoppingCartService.updateCartItemQuantity(id, cartItemRequestDto.getQuantity());
+    }
+
+    @DeleteMapping("/items/{cartItemId}")
+    public void deleteCartItem (@PathVariable Long id){
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
    public void deleteCartItem(@PathVariable Long id){
```

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        shoppingCartService.deleteCartItem(id);
+    }
+
+    @DeleteMapping("/items")
+    public void clearCart(@AuthenticationPrincipal User user) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]: remove

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    private Long bookId;
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.spring.boot.dto.cartItem;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java:10]:
```suggestion
    @NotNull
    @Positive
```

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    private Long bookId;
+    @NotNull
+    @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.Setter;
+import lombok.ToString;
+
+@Entity
+@Getter
+@Setter
+@ToString
+@Table(name = "cart_items")
+public class CartItem {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @Column(nullable = false)
+    private Integer quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
    private int quantity;
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/ShoppingCart.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+@Getter
+@Setter
+@ToString
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(name = "user_id", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/ShoppingCart.java]:
```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false)
```

File: src/main/java/mate/academy/spring/boot/repository/CartItem/CartItemRepository.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.spring.boot.repository.CartItem;
+
+import mate.academy.spring.boot.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import java.util.Set;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/repository/CartItem/CartItemRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/ShoppingCartService.java [hunk 18]
```diff
@@ -0,0 +1,18 @@
+package mate.academy.spring.boot.service;
+
+import mate.academy.spring.boot.dto.cartItem.CartItemRequestDto;
+import mate.academy.spring.boot.dto.shoppingCart.ShoppingCartDto;
+
+public interface ShoppingCartService {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/ShoppingCartService.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/UserService.java [hunk 19]
```diff
@@ -3,7 +3,10 @@
 import mate.academy.spring.boot.dto.user.UserRegistrationRequestDto;
 import mate.academy.spring.boot.dto.user.UserResponseDto;
 import mate.academy.spring.boot.exception.RegistrationException;
+import mate.academy.spring.boot.model.User;
 
 public interface UserService {
     UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException;
+
+    User findById(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/UserService.java]: remove

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    private final CartItemRepository cartItemRepository;
+    private final BookService bookService;
+    private final UserService userService;
+    private final BookMapper bookMapper;
+
+    @Override
+    public ShoppingCartDto getCartByUserId(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found with userId " + userId));
+        shoppingCart.setCartItemSet(cartItemRepository.findByShoppingCartIdWithBook(shoppingCart.getId()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto) {
+        User user = userService.findById(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public ShoppingCartDto getCartByUserId(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found with userId " + userId));
+        shoppingCart.setCartItemSet(cartItemRepository.findByShoppingCartIdWithBook(shoppingCart.getId()));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java:33]: add Transactional where is needed

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto) {
+        User user = userService.findById(userId);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByIdWithItems(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not for user " + user));
+        Book book = bookMapper.toEntity(bookService.findById(cartItemRequestDto.getBookId()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        Book book = bookRepository.findById(cartItemRequestDto.getBookId())
                .orElseThrow(()....
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found with userId " + userId));
+        shoppingCart.setCartItemSet(cartItemRepository.findByShoppingCartIdWithBook(shoppingCart.getId()));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto) {
+        User user = userService.findById(userId);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByIdWithItems(user.getId())
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+            existingCartItem.setQuantity(existingCartItem.getQuantity() + cartItemRequestDto.getQuantity());
+            cartItemRepository.save(existingCartItem);
+        } else {
+            CartItem newCartItem = new CartItem();
+            newCartItem.setBook(book);
+            newCartItem.setQuantity(cartItemRequestDto.getQuantity());
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItemSet().add(newCartItem);
+            cartItemRepository.save(newCartItem);
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        }
        shoppingCartRepository.save(shoppingCart);
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(existingCartItem.getQuantity() + cartItemRequestDto.getQuantity());
+            cartItemRepository.save(existingCartItem);
+        } else {
+            CartItem newCartItem = new CartItem();
+            newCartItem.setBook(book);
+            newCartItem.setQuantity(cartItemRequestDto.getQuantity());
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItemSet().add(newCartItem);
+            cartItemRepository.save(newCartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not for user " + user));
+        Book book = bookMapper.toEntity(bookService.findById(cartItemRequestDto.getBookId()));
+        CartItem existingCartItem = shoppingCart.getCartItemSet()
+                .stream()
+                .filter(item -> item.getBook().getId().equals(book.getId()))
+                .findFirst()
+                .orElse(null);
+        if (existingCartItem != null) {
+            existingCartItem.setQuantity(existingCartItem.getQuantity() + cartItemRequestDto.getQuantity());
+            cartItemRepository.save(existingCartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+    public ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto) {
+        User user = userService.findById(userId);
+        ShoppingCart shoppingCart = shoppingCartRepository.findByIdWithItems(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not for user " + user));
+        Book book = bookMapper.toEntity(bookService.findById(cartItemRequestDto.getBookId()));
+        CartItem existingCartItem = shoppingCart.getCartItemSet()
+                .stream()
+                .filter(item -> item.getBook().getId().equals(book.getId()))
+                .findFirst()
+                .orElse(null);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]: you can use 
```suggestion
                .ifPresentOrElse(item ->item.setQuantity(item.getQuantity() + cartItemRequestDto.getQuantity()),
                     () -> create method where you set book to cartItem and add cartItem to shop cart
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItemSet().add(newCartItem);
+            cartItemRepository.save(newCartItem);
+        }
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartDto updateCartItemQuantity(Long id, Integer quantity) {
+        CartItem cartItem = cartItemRepository.findById(id)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())...
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("CartItem not found with id " + id));
+        cartItem.setQuantity(quantity);
+        cartItemRepository.save(cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java:68]:
```suggestion
        shoppingCartRepository.save(cart);
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+        return shoppingCartMapper.toDto(cartItem.getShoppingCart());
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartMapper.toDto(cart);
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
+        CartItem cartItem = cartItemRepository.findById(id)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId())....
```

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("CartItem not found with id " + id));
+        cartItemRepository.delete(cartItem);
+    }
+
+    @Override
+    public void clearCart(Long userId) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]: remove

File: src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java [hunk 34]
```diff
@@ -29,4 +30,10 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) throws Re
         userRepository.save(user);
         return userMapper.toUserResponse(user);
     }
+
+    @Override
+    public User findById(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/impl/UserServiceImpl.java]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    private int quantity;
+
+    public Integer getQuantity() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]: why do you need this get?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+    public ShoppingCartDto updateCartItemQuantity(@AuthenticationPrincipal User user, @PathVariable("id") Long id,
+            @RequestBody @Valid CartItemRequestDto cartItemRequestDto){
+        return shoppingCartService.updateCartItemQuantity(user.getId() ,id, cartItemRequestDto.getQuantity());
+    }
+
+    @DeleteMapping("/items/{cartItemId}")
+    public void deleteCartItem(@AuthenticationPrincipal User user, @PathVariable Long id){
+        shoppingCartService.deleteCartItem(user.getId(), id);
+    }
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 37]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @NotNull
+    @Positive
+    private Long bookId;
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java:11]:
```suggestion
    private long bookId;
```
let's just use primitive, we won't need NotNull

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 38]
```diff
@@ -0,0 +1,83 @@
...
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Transactional
+    @Override
+    public void deleteCartItem(Long userId, Long id) {
+        ShoppingCart cart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("ShoppingCart not found for user " + id));
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(id, cart.getId())
+                .orElseThrow(() -> new EntityNotFoundException("CartItem not found with id " + id));
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java]: duplicated logic, better extract to private methods

File: src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java [hunk 39]
```diff
@@ -0,0 +1,83 @@
...
+
+    private void addNewCartItemToCart(ShoppingCart shoppingCart, Book book, int quantity) {
+        CartItem newCartItem = new CartItem();
+        newCartItem.setBook(book);
+        newCartItem.setQuantity(quantity);
+        newCartItem.setShoppingCart(shoppingCart);
+        shoppingCart.getCartItemSet().add(newCartItem);
+        cartItemRepository.save(newCartItem);
+    }
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/service/impl/ShoppingCartServiceImpl.java:76]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/ShoppingCartService.java [hunk 40]
```diff
@@ -0,0 +1,15 @@
...
+public interface ShoppingCartService {
+    ShoppingCartDto getCartByUserId(Long userId);
+
+    ShoppingCartDto addBookToCart(Long userId, CartItemRequestDto cartItemRequestDto);
+
+    ShoppingCartDto updateCartItemQuantity(Long userId, Long cartItemId, Integer quantity);
+
+    void deleteCartItem(Long userId, Long id);
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/service/ShoppingCartService.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/repository/shoppingCart/ShoppingCartRepository.java [hunk 41]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+import java.util.Optional;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    Optional<ShoppingCart> findByUserId(Long userId);
+
+    @Query("SELECT sc FROM ShoppingCart sc LEFT JOIN FETCH sc.cartItemSet WHERE sc.id = :id")
+    Optional<ShoppingCart> findByIdWithItems(@Param("id") Long id);
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/repository/shoppingCart/ShoppingCartRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/repository/CartItem/CartItemRepository.java [hunk 42]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.Optional;
+import java.util.Set;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Query("SELECT ci FROM CartItem ci JOIN FETCH ci.book WHERE ci.shoppingCart.id = :cartId")
+    Set<CartItem> findByShoppingCartIdWithBook(@Param("cartId") Long cartId);
+
+    Optional<CartItem> findByIdAndShoppingCartId(Long itemId, Long cartId);
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/repository/CartItem/CartItemRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/CartItem.java [hunk 43]
```diff
@@ -0,0 +1,32 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne(fetch = FetchType.LAZY, optional = false)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    private int quantity;
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/model/CartItem.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/model/ShoppingCart.java [hunk 44]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+import java.util.HashSet;
+import java.util.Set;
+
+@Entity
+@Getter
+@Setter
+@ToString
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/model/ShoppingCart.java]: Be careful with using lombok's toString on JPA entities, you have lazily loaded depended entities, which can cause lazy init exception, use @ToString(exclude = "cartItemSet")

File: src/main/java/mate/academy/spring/boot/mapper/CartItemMapper.java [hunk 45]
```diff
@@ -0,0 +1,16 @@
...
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+
+    @Mapping(target = "bookId", source = "book.id")
+    @Mapping(target = "bookTitle", source = "book.title")
+    CartItemDto toDto(CartItem cartItem);
+
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/mapper/CartItemMapper.java]:
```suggestion
```
don't leave an empty line before class ends

File: src/main/java/mate/academy/spring/boot/dto/shoppingCart/ShoppingCartDto.java [hunk 46]
```diff
@@ -0,0 +1,14 @@
+package mate.academy.spring.boot.dto.shoppingCart;
+
+import lombok.Getter;
+import lombok.Setter;
+import mate.academy.spring.boot.dto.cartItem.CartItemDto;
+import java.util.Set;
+
+@Getter
+@Setter
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/dto/shoppingCart/ShoppingCartDto.java]: it's ok to just use @Data on all DTO

**Review by okuzan — CHANGES_REQUESTED** [review]
some comments

File: src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java (hunk 36 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/spring/boot/controller/ShoppingCartController.java]: not fixed

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 47]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.spring.boot.dto.cartItem;
+
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
+    @Positive
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java:9]:
```suggestion
    @Positive
    @NotNull
```

File: src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java [hunk 48]
```diff
@@ -0,0 +1,12 @@
...
+    private long bookId;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/cartItem/CartItemRequestDto.java]:
```suggestion
    private Long bookId;
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**